### PR TITLE
Init commit for adding root sequence and rate from user

### PR DESCRIPTION
**Pull request by zmertens**
- Addresses #35 

 - I wasn't sure what to do about converting the string and vector into the Boost 64 data type since all the existing examples use `mutt`

### CODE REVIEW

File: src/include/dawg/root.h [hunk 1]
```diff
@@ -17,7 +17,18 @@ class root_model {
 	bool create(unsigned int len, const std::string &seq,  const std::vector<double> &rates) {
 		root_len = len;
 		name = "stationary";
-		do_op = &root_model::do_stat;
+		this->seq = seq;
+		this->rates = rates;
+
+		if (seq == "" && rates.empty())
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: Use seq.empty()

File: src/include/dawg/root.h [hunk 2]
```diff
@@ -44,8 +55,37 @@ class root_model {
 		}
 	}
 	
+	void do_user_seq(sequence &seq, mutt &m, const subst_model &s, const rate_model &r, residue::data_type b) const {
+		seq.resize(root_len);
+		for(sequence::iterator it=seq.begin(); it != seq.end(); ++it) {
+			it->base(static_cast<residue::data_type>(this->seq.at(0)));
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: 1) This will only every use the first character of the sequence.
2) You cannot simply cast a char to a residue. You have to translate them.

File: src/include/dawg/root.h (hunk 1 above)
Comment (reedacartwright) [inline src/include/dawg/root.h]: Use seq.empty()

File: src/include/dawg/root.h [hunk 3]
```diff
@@ -44,8 +55,37 @@ class root_model {
...
+			it->rate_cat(r(m));
+			it->branch(b);
+		}
+	}
+
+	void do_user_rates(sequence &seq, mutt &m, const subst_model &s, const rate_model &r, residue::data_type b) const {
+		seq.resize(root_len);
+		for(sequence::iterator it=seq.begin(); it != seq.end(); ++it) {
+			it->base(s(m));
+			it->rate_cat(static_cast<residue::data_type>(this->rates.back()));
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: You have to match the rate to the rate_category, casting is not going to do what you want it to.

Also this will only every use one rate from the vector.

**Review by reedacartwright — COMMENTED** [review]
Jenkins, okay to test

**Comment by reedacartwright** [comment]
Look at `dawg::residue_exchance` for code to covert between residues and strings.

**Comment by reedacartwright** [comment]
Specifying rates will require some changes to the rate_model class to match supplied rates to nearest rate categories.

**Comment by zmertens** [comment]
I think it's working now, I tested it and the output was more or less "A" with one or two SNPs sprinkled in (using the test file you sent me).

**Comment by reedacartwright** [comment]
It is not working for me, have you pushed all of your changes?

**Comment by reedacartwright** [comment]
I still cannot get your patch to work.

This input file produces correct output:
```
Tree.Tree = "(A:0.1,B:0.1);"
Root.Length = 10
```

This input file produces no output:
```
Tree.Tree = "(A:0.1,B:0.1);"
Root.Seq = "AAAAAAAAAAAA"
```

This input file produces corrupted output:
```
Tree.Tree = "(A:0.1,B:0.1);"
Root.Seq = "AAAAAAAAAAAA"
Root.Length = 10
```

Output:
```
CLUSTAL multiple sequence alignment (Created by dawg 2-current-rUnknown)

A              !TAAAAAAGG
B              !TTAAAAAAA
```

What are those `!` doing there?

I'm pretty sure codon, RNA, and AA inputs are equally broken.

**Comment by reedacartwright** [comment]
Also, do not overwrite your branch after you have created a pull request. It completely breaks the tracking and history of the pull request.

**Comment by zmertens** [comment]
Wasn't sure why the original PR wasn't viewable so I recommitted everything and then force pushed it, hopefully won't have to do that again.

**Comment by zmertens** [comment]
Should there be an error if the `root length` is not present (or a check to set the root length to the root sequence's size)?

What if the user specifies a root sequence longer then the root length?

**Comment by reedacartwright** [comment]
Okay, it seems like you are now making progress.

File: src/include/dawg/root.h [hunk 4]
```diff
@@ -43,12 +54,46 @@ class root_model {
 			it->branch(b);
 		}
 	}
-	
+
+	void do_user_seq(sequence &seq, mutt &m, const subst_model &s, const rate_model &r, residue::data_type b) const {
```
Comment (reedacartwright) [inline src/include/dawg/root.h:58]: Each segment contains a residue_exchange. Instead of creating one in the function, segment should pass its exchange to the `operator()` function as well as this one and its sisters. residue_exchange can also be modified to tell you what type of sequence needs to be created.

File: src/include/dawg/root.h [hunk 5]
```diff
@@ -15,17 +15,28 @@ namespace dawg {
 class root_model {
 public:
 	bool create(unsigned int len, const std::string &seq,  const std::vector<double> &rates) {
-		root_len = len;
+		root_len = (len == 0) ? seq.size() : len;
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: Will not work for codon models.

File: src/include/dawg/root.h [hunk 6]
```diff
@@ -43,12 +54,46 @@ class root_model {
...
+			if (s.seq_type() == residue_exchange::DNA) {
+				residue_exchange rex;
+				seq.at(i).base(rex.encode(root_seq.at(i)));
+			} else {
+
+			}
+		}
+	}
+
+	void do_user_rates(sequence &seq, mutt &m, const subst_model &s, const rate_model &r, residue::data_type b) const {
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: Let's just get rid of the option here and in the trick format to specify rates.

**Comment by reedacartwright** [comment]
I made three comments on the patch you need to address next.

File: src/include/dawg/root.h (hunk 5 above)
Comment (zmertens) [reply src/include/dawg/root.h]: Would something like this work?
```c++
if (!seq.empty() && seq.size() % 2 == 0) 
    root_len = seq.size();
else if (!seq.empty() && seq.siz() % 2 != 0)
    root_len = seq.size() / 3;
else
    root_len = len;
// otherwise error
```

Comment (reedacartwright) [reply src/include/dawg/root.h]: If `seq.empty() == false` do we even need to access `root_len`?

Comment (zmertens) [reply src/include/dawg/root.h]: In `do_user_seq` the incoming sequence gets resized to the root length, so I think that's why that one example file you sent had no output. I guess I'm not sure if we should assume that the user will specify a correctly sized root length that matches the root sequence.

Comment (reedacartwright) [reply src/include/dawg/root.h]: In what function do you want to add that code? `create` or `do_user_seq`?

Comment (zmertens) [reply src/include/dawg/root.h]: I was going to see if it could stay in `create`

Comment (reedacartwright) [reply src/include/dawg/root.h]: As written, it doesn't make sense in create since you are setting the value in `len` instead of reading from it.

`create` should just store `Root.Length` and `Root.Seq` and `do_user_seq` can rely on `residue_exchange` to convert the sequence.

**Comment by reedacartwright** [comment]
- [ ] Create `dawg::sequence residue_exchange::encode(const std::string &str)`
 - [ ] Update function signature: `root_model::create(unsigned int len, dawg:sequence seq)`
 - [ ] Update `dawg::matic::add_config_section` based on the new code.
 - [ ] Update `root_model::do_user_seq` to simply copy the stored sequence to the output seq.

File: src/include/dawg/residue.h [hunk 7]
```diff
@@ -194,9 +209,9 @@ class residue_exchange {
 		return static_cast<residue::data_type>(ret);
 	}
 
-	inline std::vector<residue> encode(const std::string &root_seq) const {
-		std::vector<residue> residues;
-		if (root_seq.size() % 2 == 0) {
+	inline sequence encode(const std::string &root_seq) const {
+		sequence residues;
+		if (type_ == DNA) {
```
Comment (reedacartwright) [inline src/include/dawg/residue.h]: type_ != CODON

File: src/include/dawg/residue.h [hunk 8]
```diff
@@ -213,8 +213,9 @@ class residue_exchange {
 		sequence residues;
 		if (type_ == DNA) {
 			for (auto i = 0; i != root_seq.size(); ++i) {
-				residues.emplace_back();
-				residues.back().base(encode(root_seq.at(i)));
+				auto base = encode(root_seq.at(i));
+				// check for errors
```
Comment (reedacartwright) [inline src/include/dawg/residue.h]: Need to fill in this check.

File: src/include/dawg/root.h [hunk 9]
```diff
@@ -56,9 +56,11 @@ class root_model {
 	}
 
 	void do_user_seq(sequence &seq, mutt &m, const subst_model &s, const rate_model &r, residue::data_type b) const {
-		std::copy(root_seq.begin(), root_seq.end(), std::back_inserter(seq));
-		// seq.at(i).rate_cat(r(m));
-		// seq.at(i).branch(b);
+		seq = root_seq;
+		for (auto 묒 = 0; 묒 != seq.size(); ++묒) {
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: Why is there a non-latin character here?

Comment (zmertens) [reply src/include/dawg/root.h]: Thought it might spice up the iteration.

Comment (reedacartwright) [reply src/include/dawg/root.h]: And make it harder to maintain code.

**Comment by reedacartwright** [comment]
Still need to add codon parsing and error checking to `encode`

**Comment by zmertens** [comment]
I tried changing the bit packing routine but I'm still getting invalid chars. I'm going to try and look closer how the `develop` handles errors and generates the codons. I feel like since we know the end-value of a codon it should be easy to match the triplet to the codon.

**Comment by zmertens** [comment]
Is there any reason for inserting an indel in the user defined root sequence?

**Comment by zmertens** [comment]
The example file that I had been using may have been bad. I tested it on `develop` branch and I got output that would suggest the input format was bad (wrong alphabet of letters). I'll attach the file and output for reference.

[models_out_from_specify_root_sequence.txt](https://github.com/reedacartwright/dawg/files/1455148/models.txt)
[models_out_from_develop.txt](https://github.com/reedacartwright/dawg/files/1455153/models_out_from_develop.txt)

I can confirm that the `multiple_models` example is consistent when specifying a root sequence and on the develop and that uses a codon model.

**Comment by zmertens** [comment]
Updated. Kind of forgot how long this PR has been open for, so I didn't need to reopen a new one.

File: src/include/dawg/residue.h [hunk 10]
```diff
@@ -66,28 +96,36 @@ class residue {
...
 
+	explicit residue_exchange(int m=DNA) { model(m,0,0,false,false,false); }
+
 	inline bool model(unsigned int type, unsigned int code, bool rna,
 			bool lowercase, bool markins, bool keepempty) {
+
 		static constexpr char sIns[] = "-+";
 		// table for going from base->char
 		// TODO: Allow codons to be translated into aa
+		// table is 63 wide and 30 down
```
Comment (reedacartwright) [inline src/include/dawg/residue.h]: 64 wide, right?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
+language: c++
```
Comment (reedacartwright) [inline .travis.yml]: Remove the travis file from the patch

File: .editorconfig [hunk 12]
```diff
@@ -0,0 +1,29 @@
+language: c++
```
Comment (reedacartwright) [inline .editorconfig]: Put this file into separate pull request.

File: dawg_diff.sh [hunk 13]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/bash
```
Comment (reedacartwright) [inline dawg_diff.sh]: Why is this part of the pull request?

File: examples/aa.dawg [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+Sim.Reps = 10
+Sim.Seed = 42
+Root.Seg = 0
+
+Subst.Model = "aagtr"
+Subst.Freqs  = 0.2, 0.3, 0.3, 0.2
+Subst.Params = 2.0, 1.0, 3.0, 1.0, 1.0, 1.0
+Subst.Rate.Model = "ZERO"
+
+Subst.Model = "equ"
```
Comment (reedacartwright) [inline examples/aa.dawg]: You are overwriting the previous model here. Did you realize that?

File: src/include/dawg/root.h [hunk 15]
```diff
@@ -5,27 +5,38 @@
...
 #include <dawg/rate.h>
-#include <dawg/sequence.h>
+#include <dawg/residue.h>
 
 namespace dawg {
 
 class root_model {
 public:
-	bool create(unsigned int len, const std::string &seq,  const std::vector<double> &rates) {
+	bool create(unsigned int len, sequence &&root_seq) {
```
Comment (reedacartwright) [inline src/include/dawg/root.h]: Do not use an rvalue reference in arguments. If you want to move something use copy by value in the arguments so the caller can decide whether they want to move it or not.

**Review by reedacartwright — CHANGES_REQUESTED** [review]

File: examples/aa.dawg (hunk 14 above)
Comment (zmertens) [reply examples/aa.dawg]: Should they be part of separate segments?

Comment (reedacartwright) [reply examples/aa.dawg]: Separate segments or separate files.

Comment (zmertens) [reply examples/aa.dawg]: ~~I'm having trouble setting up a `aagtr` model. It says it needs 20 frequencies and 190 parameters~~

I got the dna, codon, and aa examples working. When I first put it together I didn't realize that I could use dawg's inheritance model to just specify the `subst.freq` and `subst.params` once.

File: src/include/dawg/root.h (hunk 15 above)
Comment (zmertens) [reply src/include/dawg/root.h]: I should note that after modifying the rvalue reference it won't compile because in `matic.cpp` it won't bind to a temporary value (without the move), but I fixed it by changing the caller to not use a temporary.

**Review by reedacartwright — APPROVED** [review]
